support generalized eigenvalue problems

fdkong opened the pull request:
Enhance `NonlinearEigenSystem` and `EigenProblem`  to support generalized eigenvalue problems.

Added an example to show the usage.

Refs #7398

fdkong committed:
Added SlepcSupport.{C,h} to handle slepc related

options. This will be used to setup slepc options.

Refs #7398

fdkong committed:
Added interfaces in EigenProblem and

NonlinearEigenSystem to introduce generalized eigenvalue problems

Refs #7398

fdkong committed:
Added a flag to Kernel base

to indicate if it will be used an eigen kernel.

Refs #7398

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,52 @@
+  {
+    InputParameters params = emptyInputParameters();
+
+    MooseEnum eigenvalue_problem_type("HEP NHEP GHEP GNHEP GHIEP");
```
does missing the second argument means that there is no default value for this moose enum and the later parameter is required?

fdkong replied:
> does missing the second argument means that there is no default value for this moose enum and the later parameter is required?

We do not set the default parameter here. If users do not  setup anything in the input file, we have a default parameter setup from them in the solver.  The reason I do not like a default setup here is that I want to use` params.isParamValid("")`.

YaqiWang commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,52 @@
+    InputParameters params = emptyInputParameters();
+
+    MooseEnum eigenvalue_problem_type("HEP NHEP GHEP GNHEP GHIEP");
+    params.addParam<MooseEnum>   ("eigenvalue_problem_type", eigenvalue_problem_type,
```
extra space between `<MooseEnum>` and `(`.

YaqiWang commented on framework/src/base/EigenProblem.C:
```diff
@@ -17,19 +17,29 @@
   params.addParam<unsigned int>("n_basis_vectors", 3, "The dimension of the nullspace");
+  params.addParam<bool>("generalized_eigenvalue_problem", false, "if we are trying to solve a generalized eigenvalue problem");
+#if LIBMESH_HAVE_SLEPC
+  params += Moose::SlepcSupport::getSlepcValidParams();
```
`eigenvalue_problem_type` is never used?

YaqiWang replied:
And the previous parameter `generalized_eigenvalue_problem` looks redundant with the SLEPc parameter `eigenvalue_problem_type`.

fdkong replied:
Removed

YaqiWang commented on framework/src/base/NonlinearEigenSystem.C:
```diff
@@ -30,10 +30,20 @@ namespace Moose {
+    if (eigen_system.matrix_B)
+      p->computeJacobian(*eigen_system.current_local_solution.get(), *eigen_system.matrix_B);
+    else
+      mooseError("It is a generalized eigenvalue problem but matrix B is empty \n");
```
So at this point we have to explicitly assemble B and no B-matrix-free is supported?

fdkong replied:
No matrix-free at this point. Matrix-free will be added in another PR

YaqiWang commented on framework/include/kernels/KernelBase.h:
```diff
@@ -99,6 +99,9 @@ class KernelBase :
   SubProblem & subProblem();
 
+  /// if is it an eigen kernel
+  virtual isEigenKernel() { return _eigen_kernel; }
```
missing return type `bool`?

YaqiWang commented on framework/src/kernels/KernelBase.C:
```diff
@@ -43,6 +43,8 @@ InputParameters validParams<KernelBase>()
 
   params.declareControllable("enable");
+
+  params.addParam<bool>   ("eigen_kernel", false,"Whether or not this kernel will be used as an eigen kernel");
```
extra space and missing space after `false,`.

YaqiWang commented on framework/include/kernels/MassKernel.h:
```diff
@@ -0,0 +1,36 @@
+/****************************************************************/
```
this kernel is identical with `Reaction`.

fdkong replied:
I did not know this, and I will remove this kernel.

YaqiWang commented on framework/include/kernels/KernelBase.h:
```diff
@@ -100,7 +100,7 @@ class KernelBase :
 
   /// if is it an eigen kernel
-  virtual isEigenKernel() { return _eigen_kernel; }
+  virtual bool isEigenKernel() { return _eigen_kernel; }
```
this should be amended in the previous commit...

YaqiWang commented on framework/src/base/NonlinearSystemBase.C:
```diff
@@ -1699,6 +1699,7 @@ NonlinearSystemBase::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
     default:
     case Moose::COUPLING_CUSTOM:
       {
+        std::cout<<" Moose::COUPLING_CUSTOM "<<std::endl;  
```
std::cout?

YaqiWang commented on framework/src/base/EigenProblem.C:
```diff
@@ -17,19 +17,29 @@
 {
   InputParameters params = validParams<FEProblemBase>();
   params.addParam<unsigned int>("n_eigen_pairs", 1, "The dimension of the nullspace");
   params.addParam<unsigned int>("n_basis_vectors", 3, "The dimension of the nullspace");
```
To me this parameter should be hidden to the user or at least group it into 'advanced'. Also give the recommendation on the value of this param in its doc string.

YaqiWang commented on test/tests/problems/eigen_problem/gipm.i:
```diff
@@ -0,0 +1,126 @@
+#Its inverse will be 0.49950700634518.
+
+[Variables]
+  active = 'u'
```
no need for this active.

YaqiWang commented on test/tests/problems/eigen_problem/gipm.i:
```diff
@@ -0,0 +1,126 @@
+[]
+
+[Kernels]
+  active = 'diff rea rhs'
```
no need

YaqiWang commented on test/tests/problems/eigen_problem/gipm.i:
```diff
@@ -0,0 +1,126 @@
+[]
+
+[BCs]
+  active = 'homogeneous'
```
no need

YaqiWang commented on test/tests/problems/eigen_problem/gipm.i:
```diff
@@ -0,0 +1,126 @@
+[Executioner]
+  type = Steady
+  solve_type = 'NEWTON'
+  petsc_options = '-eps_smallest_magnitude'
```
this two SLEPc parameter can be changed to something like what we did for, for example, `line_search` because they are so essential.

permcody replied:
@YaqiWang - This is a good idea. I'm not sure exactly what you want but we want the interface to be nice and approachable for users. @fdkong could probably use feedback on his design decisions. He's a hardened PETSc user so he isn't personally bothered by seeing lots of PETSc options appearing raw in his input file.

YaqiWang commented on framework/include/utils/SlepcSupport.h:
```diff
@@ -15,7 +15,7 @@
 #ifndef SLEPCSUPPORT_H
```
this commit should not exist and changes here should be amendment in the previous commits.

fdkong committed:
Added a test for generalized problems

Refs #7398

fdkong committed:
Works for generalized eigenvalue problems

Refs #7398

fdkong committed:
Wrap eigen solver type so that users

directly use eigen_solve_type to setup an eigenvalue solver

Refs #7398

moosebuild commented:
Job `Test intel:linux-intel` on 3478f8d : invalidated by @fdkong

timeout

fdkong committed:
Wrap eigen problem type

Use eigen_problem_type to setup the type of eigenproblem

Refs #7398

fdkong committed:
Added an input paramter "which_eigen_pairs"

to allow users to specify eigenvalues of interest

Refs #7398

moosebuild commented:
Job `Test intel:linux-intel` on cf9c547 : invalidated by @fdkong

timeout

moosebuild commented:
Job `Test intel:linux-intel` on cf9c547 : invalidated by @fdkong

Try again

moosebuild commented:
Job `Test intel:linux-intel` on cf9c547 : invalidated by @fdkong

timeout

permcody reviewed (changes requested):
Pretty close, please address these comments.

YaqiWang commented:
@fdkong I do not know why lots of my comments show outdated although they are not. Please also address them along with @permcody 's. Thanks.

fdkong commented:
> @fdkong I do not know why lots of my comments show outdated although they are not. Please also address them along with @permcody 's. Thanks.

@YaqiWang, I addressed all of your comments. Which one I missed?

fdkong commented on framework/src/base/ComputeResidualThread.C:
```diff
@@ -103,6 +103,17 @@ ComputeResidualThread::onElement(const Elem *elem)
     break;
+
+  case Moose::KT_EIGEN:
+    warehouse = &_nl.getEigenKernelWarehouse();
```
I am not going to change the old ones. Actually, it is not necessary to store a member for this because it is used once only. That is, we access this warehouse at line 108, and then we do not access again.

fdkong replied:
If you prefer to the old styles, I could change these.

fdkong commented on framework/src/base/EigenProblem.C:
```diff
@@ -46,6 +55,12 @@ EigenProblem::EigenProblem(const InputParameters & parameters) :
+
+  _eq.parameters.set<EigenProblem *>("_eigen_problem") = this;
+
+  Moose::SlepcSupport::storeSlepcEigenProblemOptions(*this, _pars);
```
NO. We already have it at line 44.

fdkong commented on framework/src/base/NonlinearEigenSystem.C:
```diff
@@ -135,4 +145,35 @@ NonlinearEigenSystem::getNthConvergedEigenvalue(dof_id_type n)
 
+
+void
+NonlinearEigenSystem::addKernel(const std::string & kernel_name, const std::string & name, InputParameters parameters)
```
Will do

fdkong commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,249 @@
+    // can not use solve_type for eigenvalue problems
+    if (params.isParamValid("solve_type"))
+    {
+      mooseError("Can not use solve_type for eigenvalue problems, please use eigen_solve_type instead \n");
```
Will do

moosebuild commented:
Job `Test intel:linux-intel` on 0abb6f8 : invalidated by @fdkong

timeout

fdkong commented:
@permcody, @YaqiWang  more suggestions?

permcody commented on framework/include/kernels/KernelBase.h:
```diff
@@ -99,6 +99,9 @@ class KernelBase :
   /// Returns a reference to the SubProblem for which this Kernel is active
   SubProblem & subProblem();
 
+  /// if is it an eigen kernel
```
Captain Obvious

permcody commented on framework/include/kernels/KernelBase.h:
```diff
@@ -174,6 +177,9 @@ class KernelBase :
   std::vector<MooseVariable*> _diag_save_in;
   std::vector<AuxVariableName> _diag_save_in_strings;
+
+  /// Whether or not this kernel will be used as an eigen kernel
```
Captain Obvious

permcody commented on framework/include/outputs/OutputWarehouse.h:
```diff
@@ -189,6 +189,43 @@ class OutputWarehouse
    */
   void bufferConsoleOutputsBeforeConstruction(bool buffer) { _buffer_action_console_outputs = buffer; }
 
+public:
```
This label is unecessary. We are already in the public declaration region.

permcody commented on framework/include/outputs/OutputWarehouse.h:
```diff
@@ -189,6 +189,43 @@ class OutputWarehouse
+   * Calls the initialSetup function for each of the output objects
+   * @see FEProblemBase::initialSetup()
+   */
+  void initialSetup();
```
This is a big change. We are moving all of these setup methods from private to public?

fdkong replied:
It would be nice if we could claim these setups as public, otherwise a class has to be claimed as friend of OutputWarehouse if the class want to access these routines.  Originally, I want to implement a SlepcOutput which need to access these setups, but later on I have to discard the implementation for some reason. 

I will change these routines back to private. In the future, we will claim them as public again when necessary.

permcody replied:
If making them public avoids "friend", then that's the right thing to do. I just wanted to understand why.

friedmud replied:
Not necessarily.  There is nothing inherently wrong with `friend`.  It all depends on the context.  Making things `public` means that anyone is free to go at that function... `friend` can be used to allow internal objects to work together without exposing functionality to the whole world.

In our case with MOOSE I often use `friend` so that I can allow MOOSE internal classes more access to each other while shielding our users (or shielding ourselves from our users!) from using that capability.

permcody commented on framework/include/utils/MooseTypes.h:
```diff
@@ -166,6 +166,8 @@ enum KernelType
   KT_TIME = 0,
   KT_NONTIME = 1,
+  KT_NONEIGEN = 2,
+  KT_EIGEN  = 3,
```
Do we use any of the enumeration constants directly? Maybe we should just drop them.

fdkong replied:
We use them in a few places: `FEProblemBase`, `ComputeResidualThread` and `ComputeJacobianThread`.

Following the existing implementation for time kernels, I have to handle the eigenvalue kernels like this way. Any better way are welcome.

permcody replied:
I was talking about the constants (0, 1, 2, 3) not the enumerations. I'm just saying that we probably don't need the numbers.

fdkong replied:
I followed  `time kernels` to add constants. I will remove constants. I have been trying my best to avoid changing the exiting code.

permcody replied:
I appreciate that in general. However sometimes we run into bad designs as we add new capability. I think it's worth refactoring those designs in additional commits as we go. It helps make things more maintainable and I'm all for doing that.

andrsd replied:
`KT_TIME` and `KT_NONTIME` is used by time integration system. The constants _have_ to stay, because they are used in `Assembly` to index into local residuals for time and non-time contribution for the residual. They are not used only for kernel type...

permcody commented on framework/src/base/FEProblemBase.C:
```diff
@@ -3612,13 +3612,18 @@ FEProblemBase::computeResidualType(const NumericVector<Number>& soln, NumericVec
 void
 FEProblemBase::computeJacobian(const NumericVector<Number> & soln, SparseMatrix<Number> & jacobian)
 {
+  computeJacobian(soln, jacobian, Moose::KT_ALL);
```
Can we just drop this routine instead and provide a default param?

fdkong replied:
For `FEProblem`, most of people will use `computeJacobian(soln, jacobian, Moose::KT_ALL);` But for `EigenProblem`, we have two routines:  ` computeJacobian(soln, jacobian, Moose::KT_NONEIGEN )`, and  `computeJacobian(soln, jacobian, Moose::KT_EIGEN )`.

permcody commented on framework/src/base/NonlinearEigenSystem.C:
```diff
@@ -135,4 +145,14 @@ NonlinearEigenSystem::getNthConvergedEigenvalue(dof_id_type n)
 }
 
+void
+NonlinearEigenSystem::addKernelToExtraWarehouse(MooseSharedPointer<KernelBase> kernel, THREAD_ID tid)
```
I'm not sure if I like the name of this routine "...ExtraWarehouse". Are the EigenKernels in the KernelWarehouse AND one of these other two warehouses too?

fdkong replied:
This routine handles the EigenKernels only. I say `ExtraWarehouse` means if we need to handle other kernels which can not be handled by `NonlinearSystemBase`.  `All kernels`, and "Time kernels" are handled in `NonlinearSystemBase`.  My motivation is to make the new implementation consistent with the existing MOOSE.

permcody replied:
How about the name `addEigenKernels()` then?

fdkong replied:
Yes, right, only `EigenProblem` uses  `ExtraWarehouse` now, but in the future, we possibly  create `XXXProblem` to add `XXX kernels`. It would be much better if we could have a generic name.

permcody commented on framework/src/base/NonlinearSystemBase.C:
```diff
@@ -278,6 +278,9 @@ NonlinearSystemBase::addKernel(const std::string & kernel_name, const std::strin
     else
       _non_time_kernels.addObject(kernel, tid);
+
+    // store kernels to warehouse, if necessary
```
This comment doesn't seem right. It looks like we are always adding kernels in the `addExtraWarehouse()` routine. We have a lot of different warehouses for kernels now. It looks like at least 5 different warehouses. I wonder if we should consolidate this a bit and just add more methods to the warehouse for proper retrieval.

permcody replied:
This comment should say which kernels are added here, just Eigen?

fdkong replied:
`addExtraWarehouse` is doing nothing in `NonlinearSystemBase`, and its behavior will be implemented in the subclass. 

Adding more methods to warehouse may be  a good idea.

permcody reviewed (changes requested):
A few more minor comments. The one thing I'm not crazy about is the explosion of warehouses. I believe we have 5 kernel warehouses now and I don't see that it buys us anything at all. We still have to retrieve the kernels out of each one specifically. It's not like we have a generic routine that takes a "KernelWarehouseBase" or something. I know this isn't your design but this PR exacerbates the bad design that's already here.

permcody commented:
Oh I left a few "Captain Obvious" comments. @andrsd started that and it's a good thing. We don't want comments that don't add value. If you have a routine that has a descriptive name, you don't need to repeat that name as a comment for what the routine does. We want comments that a developer might actually read and get value from. Notes about why the design is the way it is, or a summary of a complex algorithm, etc.

andrsd commented:
One comment in general. As I said `KT_TIME` and `KT_NONTIME` is used to determine the kernel type. Based on this we grab the right kernels from the right warehouse when forming residuals/jacobians. However, it is also used to determine residual type. We have 2 vectors, one for time contributions and one for non-time ones, in `Assembly`.  If we want to have `KT_EIGEN` and `KT_NONEIGEN` to determine the kernel type, we should most likely introduce `RT_XYZ` (for residual type). Otherwise the interface on `Assembly` will be confusing, since right now it has things like `addResidual(..., KernelType ...)` and if somebody calls it with `KT_EIGEN` or `KT_NONEIGEN`, they get out of array access, since we allocate 2 vectors only (not 4).

fdkong commented:
> One comment in general. As I said KT_TIME and KT_NONTIME is used to determine the kernel type. Based on this we grab the right kernels from the right warehouse when forming residuals/jacobians. However, it is also used to determine residual type. We have 2 vectors, one for time contributions and one for non-time ones, in Assembly. If we want to have KT_EIGEN and KT_NONEIGEN to determine the kernel type, we should most likely introduce RT_XYZ (for residual type). Otherwise the interface on Assembly will be confusing, since right now it has things like addResidual(..., KernelType ...) and if somebody calls it with KT_EIGEN or KT_NONEIGEN, they get out of array access, since we allocate 2 vectors only (not 4).

Thanks a lot, @andrsd. Now I know why my matrix-free version does not work.  For the matrix-free version of eigenvalue solvers, we do need to assemble two different types of residual, namely, `Eigen_residual `and `NonEigen_residual `. 

Could you possibly specify what do you mean by  `RT_XYZ`?

andrsd commented:
> Could you possibly specify what do you mean by RT_XYZ

I was thinking `RT_TIME` and `RT_NONTIME` to label residuals holding the contribution from time kernels (`KT_TIME`) and non-time kernels (`KT_NONTIME`).

You will probably want to use those for your `eigen` and `noneigen` residuals, but then the name would be misleading. And `RT_1` and `RT_2` is not really descriptive.

If you cannot come up with better names, we could also abandon the Enum approach for the residuals and change the API to take an integer, which would be the index, and use `#define` which are _evil_ .

So pick your poison... ;-)

fdkong commented:
> So pick your poison... ;-)

Thanks, @andrsd.  What I need to do is so simple, just to take my poison...:-)

fdkong commented:
@andrsd, @permcody.

I took a look into` Assembly.C`, and there are some functions such as `addResidual(NumericVector<Number> & residual, Moose::KernelType type)`. The functions use `KernelType` as an input parameter.  These functions are called from `FEProblemBase` or a few other places statically when compute the residual.  I guess the main purpose is to compute the time residual and non-time residual at the same time  (using a single call of` computeResidual()`) for time kernels and non-time kernels, respectively. 

I am possibly NOT going to change anything here because:

(1) Make the moose more complicated not simpler. If we are going to solve a time-dependent eigenvalue problem, it is highly nontrivial to control the kernel type because we will have combinations like: `KT_TIME_EIGEN`, `KT_TIME_NONEIGEN`,  `KT_NONTIME_NONEIGEN`, and `KT_NONTIME_EIGEN`. Refactoring `Assembly.C`, `FEProblemBase` and so on makes the code twisted  more.

(2) It is NOT necessary to add the residual type (RT) to make the eigensolver work.  What I need to do is to follow what I have done for `computeJacobian()` to make the matrix-free version of the eigenvalue solver work. For the matrix-based eigenvalue solver, we call `computeJacobian()` twice to get `matrix A` and `matrix B`. For the matrix-free version, we could also call` computeResidual()` twice to get the residual for A and the residual for B. 

(3) I am NOT going to create a `EIgenKernel` class so everybody derives from there to avoid duplicating the code. Sometimes the matrix A and the matrix B may come from the same kernels (operators).

andrsd commented:
Not having `EigenKernel` class sounds like a plan ;-) Go ahead...

fdkong commented:
> Not having EigenKernel class sounds like a plan ;-) Go ahead...

Thanks, @andrsd. The matrix-based eigenvalue solvers (for standard and generalized ) have been implemented in this PR without using `EigenKernel`

I am going to implement the matrix-free version in another PR because it will take a while. The libMesh `eigen_system` does not support matrix-free, and I have to add something in libMesh.

permcody reviewed (approved):
This looks OK to me. I'll have @andrsd review and merge. Thanks!

andrsd commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,257 @@
+  {
+    InputParameters params = emptyInputParameters();
+
+    // We are solving a Non-Hermitian eigenvalue problem by default
```
I do not think this is true. Non-hermitian is the _second_ value in the enum and the enum does not provide "NHEP" as a default value.

fdkong replied:
The default value  is set in the class: `SolverParams`. If users do not implicitly use` eigen_problem_type` in the input file, then there is no any thing in `InputParameters`.

andrsd replied:
Why could not you use the default value in this enum then?

andrsd replied:
I see now. There is a disconnect between this `MooseEnum` and `EigenProblemType`. While your comment refers to `EPT_NHEP` which is in `SolverParams`, but sits right before this `MooseEnum`, which is confusing...

Also, if I understand correctly `slepcSetOptions` will force the options/type/etc. to be whatever the values are in the input parameters. So, no matter what the values are in the default c-tor of `SolverParams`, they get overwritten by that call. Correct?

fdkong replied:
`slepcSetOptions` actually forces options/type/etc to whatever we have in `SolverParams`. `InputParameters` goes to overwrite the values in `SolverParams` only if users implicitly add the corresponding parameters  in the input file.  If users do not do anything in the input file, `params.isParamValid("eigen_problem_type")` and `(params.isParamValid("which_eigen_pairs"))` should be false, and then we will not change values in `SolverParams`

andrsd replied:
Ok. If there was a default value set in those `InputParameters`, then you would not need the `isParamValid()` check (it would be always true). And then you would have just one place with defaults. Just something to be aware of...

andrsd commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,257 @@
+                               "GHIEP: Generalized indefinite Hermitian ");
+
+
+    // Which eigenvalues are interest
```
"Which eigenvalues are we interested in" or "Which eigenvalues are of interest"...

andrsd commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,257 @@
+  void
+  setEigenProblemOptions(SolverParams & solver_params)
+  {
+    // set eigen problem types
```
1. you are setting _options_ not a type
2. you do not need this cpt. Obvious here, the function name is pretty descriptive.

andrsd commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,257 @@
+      // which eigen pairs
+      const std::string & which_eigen_pairs = params.get<MooseEnum>("which_eigen_pairs");
+      eigen_problem.solverParams()._which_eigen_pairs = Moose::stringToEnum<Moose::WhichEigenPairs>(which_eigen_pairs);
+    }
```
These 4 comments do not really help anything...

andrsd commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,257 @@
+    // eigen solve type
+    if (params.isParamValid("eigen_solve_type"))
+    {
+      // Extract the eigen solve type
```
Again, 2 comments with questionable value...

andrsd commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,257 @@
+  void
+  setWhichEigenPairsOptions(SolverParams & solver_params)
+  {
+    // set eigen problem types
```
Copy and paste comment, this is not true. And really not needed...

andrsd commented on test/tests/problems/eigen_problem/gipm.i:
```diff
@@ -0,0 +1,117 @@
+
+[Variables]
+  [./u]
+    # second order is way better than first order
```
So why are you using `FIRST` order? We do support SECOND order lagrange elements...

fdkong replied:
I should not copy this comment from Yaqi. I have no idea whether the SECOND is better than the FIRST order or not. I will remove this comment to avoid confusing people.

YaqiWang replied:
Changing to second order, you will see the eigenvalue is much closer to the analytical one.

permcody commented:
I personally take a pretty hard stance against "friend" myself. I think
it's pretty bad in most cases as it usually means you have design problem.
Yes there are a few good cases for it but I'll add that our practical use
of it in MOOSE is almost always wrong. Specifically, we never grant friend
access to specific methods of a function. We always grant access to all
functions of a class. This is bad because it completely breaks
encapsulation.



On Mon, Jan 23, 2017 at 1:01 PM Derek Gaston <notifications@github.com>
wrote:

> *@friedmud* commented on this pull request.
> ------------------------------
>
> In framework/include/outputs/OutputWarehouse.h
> <https://github.com/idaholab/moose/pull/8277>:
>
> > @@ -189,6 +189,43 @@ class OutputWarehouse
>     */
>    void bufferConsoleOutputsBeforeConstruction(bool buffer) { _buffer_action_console_outputs = buffer; }
>
> +public:
> +  /**
> +   * Calls the initialSetup function for each of the output objects
> +   * @see FEProblemBase::initialSetup()
> +   */
> +  void initialSetup();
>
> Not necessarily. There is nothing inherently wrong with friend. It all
> depends on the context. Making things public means that anyone is free to
> go at that function... friend can be used to allow internal objects to
> work together without exposing functionality to the whole world.
>
> In our case with MOOSE I often use friend so that I can allow MOOSE
> internal classes more access to each other while shielding our users (or
> shielding ourselves from our users!) from using that capability.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8277>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XICFfLmaFsuPTtijlIlQf1iQuCecgks5rVQcugaJpZM4LcHCD>
> .
>

fdkong commented on framework/include/utils/SlepcSupport.h:
```diff
@@ -0,0 +1,48 @@
+
+#ifdef LIBMESH_HAVE_SLEPC
+
+// MOOSE includes
```
I though it is part of the MOOSE standard :) because I see everywhere.

andrsd replied:
No, it is cpt. Obvious copied around way too much...

friedmud commented:
I still maintain that it doesn't always mean you have a design problem.
For instance, in this case we don't necessarily want all of those functions
to be public... there's no reason for _most_ classes to call them... but we
might need to provide an interface to another class that's intimately
linked to this one.

In a time of "you allow it" I think we should hide almost everything we
can...

Derek

On Mon, Jan 23, 2017 at 3:16 PM Cody Permann <notifications@github.com>
wrote:

> I personally take a pretty hard stance against "friend" myself. I think
> it's pretty bad in most cases as it usually means you have design problem.
> Yes there are a few good cases for it but I'll add that our practical use
> of it in MOOSE is almost always wrong. Specifically, we never grant friend
> access to specific methods of a function. We always grant access to all
> functions of a class. This is bad because it completely breaks
> encapsulation.
>
>
>
> On Mon, Jan 23, 2017 at 1:01 PM Derek Gaston <notifications@github.com>
> wrote:
>
> > *@friedmud* commented on this pull request.
> > ------------------------------
> >
> > In framework/include/outputs/OutputWarehouse.h
> > <https://github.com/idaholab/moose/pull/8277>:
> >
> > > @@ -189,6 +189,43 @@ class OutputWarehouse
> > */
> > void bufferConsoleOutputsBeforeConstruction(bool buffer) {
> _buffer_action_console_outputs = buffer; }
> >
> > +public:
> > + /**
> > + * Calls the initialSetup function for each of the output objects
> > + * @see FEProblemBase::initialSetup()
> > + */
> > + void initialSetup();
> >
> > Not necessarily. There is nothing inherently wrong with friend. It all
> > depends on the context. Making things public means that anyone is free to
> > go at that function... friend can be used to allow internal objects to
> > work together without exposing functionality to the whole world.
> >
> > In our case with MOOSE I often use friend so that I can allow MOOSE
> > internal classes more access to each other while shielding our users (or
> > shielding ourselves from our users!) from using that capability.
> >
> > —
> > You are receiving this because you were mentioned.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/8277>, or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AC5XICFfLmaFsuPTtijlIlQf1iQuCecgks5rVQcugaJpZM4LcHCD
> >
> > .
> >
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8277#issuecomment-274604038>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMVEQogx61XFHgzZjr9QktgDZg1X8ks5rVQqQgaJpZM4LcHCD>
> .
>

fdkong commented:
@andrsd, any more comments? Thanks!

andrsd commented on framework/src/base/ComputeJacobianThread.C:
```diff
@@ -59,9 +61,39 @@ ComputeJacobianThread::~ComputeJacobianThread()
+  }
+
+  if (!warehouse)
+    mooseError("Warehouse is NULL \n");
```
I do not know if you had this in your previous version, but the only way how `warehouse == NULL` here is that somebody will add a new `KernelType` and implements and "empty" branch in the above switch statement. Note that with new `KernelType` and no branch for it, the code will stop on that `mooseError` in the `default` branch of the switch above...

andrsd commented on framework/src/base/ComputeResidualThread.C:
```diff
@@ -89,22 +85,36 @@ ComputeResidualThread::onElement(const Elem *elem)
 
+  if (!warehouse)
+    mooseError("Warehouse is NULL \n");
+
```
Similarly here...

andrsd commented on framework/src/base/NonlinearSystemBase.C:
```diff
@@ -1847,7 +1849,10 @@ NonlinearSystemBase::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
     } // end loop over boundary nodes
 
     // Set the cached NodalBC values in the Jacobian matrix
```
This comment belonged to the line `_fe_problem.assembly(0).setCachedJacobianContributions(jacobian);` which is now in the else branch.  If `kernel_type == Moose::KT_EIGEN` this comment would be a lie. Why are we dropping the cached values when `kernel_type == Moose::KT_EIGEN`? Good opportunity for an explanation in a comment...

andrsd commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,242 @@
+                               "TARGET_MAGNITUDE: target magnitude "
+                               "TARGET_REAL: target real "
+                               "TARGET_IMAGINARY: target imaginary "
+                               "ALL_EIGENVALUES: all eigenvalues ");
```
The doco string should read something like: "Which eigenvalue pair to obtain from the solution". The values in the `MooseEnum` are quite self-explanatory now...

andrsd commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,242 @@
+                               "GEN_HERMITIAN: Gerneralized Hermitian "
+                               "GEN_NON_HERMITIAN: Generalized Non-Hermitian "
+                               "GEN_INDEFINITE: Generalized indefinite Hermitian "
+                               "POS_GEN_NON_HERMITIAN: Generalized Non-Hermitian with positive (semi-)definite B");
```
Doco string should _also_ mention something like "Type of the eigenvalue problem we are solving", then it can elaborate on what the values are...

andrsd commented on framework/src/utils/SlepcSupport.C:
```diff
@@ -0,0 +1,242 @@
+    params.addParam<MooseEnum>("eigen_problem_type", eigen_problem_type,
+                               "HERMITIAN: Hermitian "
+                               "NON_HERMITIAN: Non-Hermitian "
+                               "GEN_HERMITIAN: Gerneralized Hermitian "
```
`s/Gerneralized/Generalized/`

fdkong commented:
@andrsd, tests passed :)

andrsd commented on framework/src/base/NonlinearEigenSystem.C:
```diff
@@ -152,4 +154,20 @@ NonlinearEigenSystem::addEigenKernels(MooseSharedPointer<KernelBase> kernel, THR
+
+  if (!nbc)
+    mooseError("Invalid NodalBC for eigenvalue problems, please use homogeneous Dirichlet.");
+}
```
You want to say "boundary condition" in these error messages, "boundary" means the physical "edge" of the mesh.

andrsd commented on framework/src/base/NonlinearSystemBase.C:
```diff
@@ -333,6 +333,8 @@ NonlinearSystemBase::addBoundaryCondition(const std::string & bc_name, const std
   _vars[tid].addBoundaryVar(boundary_ids, &bc->variable());
 
+  checkEigenBoundaryConditionConsistency(bc, parameters);
+
```
I do not think this is correct.

1. if I had a normal FEProblem and I added a homogeneous Dirichlet BC, I would get an error that it not possible, but it is perfectly fine. Can you verify that on your branch? I do not think we have a test with homogeneous BCs.
2. doing this check every time we add BC is probably ok, but I though we would do it in `checkProblemIntegrity` that is in `FEProblemBase`. That method should become virtual and it should call something like `_nl.checkIntegrity()` (a new method on our systems) where you would go over the objects from the warehouses and did your check.There is already a integrity checking action in MOOSE, so we should use that when ever possible...

andrsd commented on framework/src/base/NonlinearSystemBase.C:
```diff
@@ -1850,11 +1852,11 @@ NonlinearSystemBase::computeJacobianInternal(SparseMatrix<Number> & jacobian, Mo
+      _fe_problem.assembly(0).zeroCachedJacobianContributions(jacobian);
+    // Set the cached NodalBC values in the Jacobian matrix
+    else
+      _fe_problem.assembly(0).setCachedJacobianContributions(jacobian);
```
Is there a change you can move this into 98ac4f8e5a32fd7ac0336733107e6d6d7e2b8cbd where it belongs?

fdkong committed:
check boundary conditions for eigenvalue problem

For eigenvalue problems (including standard and generalized), inhomogeneous (Dirichlet or Neumann)
boundary conditions are  not allowed.

Refs #7398

fdkong committed:
Added tests for checkIntegrity

Refs #7398

fdkong commented:
@andrsd, more :)

andrsd commented on framework/src/base/EigenProblem.C:
```diff
@@ -124,6 +124,15 @@ EigenProblem::computeJacobian(const NumericVector<Number> & soln, SparseMatrix<N
+EigenProblem::checkProblemIntegrity()
+{
+  FEProblemBase::checkProblemIntegrity();
+#if LIBMESH_HAVE_SLEPC
```
Why do we need this `#ifdef`? Is the integrity check SLEPc dependent?

fdkong replied:
If there does not exist SLEPc,  `NonlinearEigenSystem` is a shell class (the class has the constructor only). ` _nl_eigen ` is a pointer of the instance of `NonlinearEigenSystem`. ` _nl_eigen` has no function member `checkIntegrity()` anymore if SLEPc is not available.

andrsd replied:
We can keep this in for right now, but we need to fix the `#ifdef` business for SLEPc, because it is getting out of hands. A concept that worked for me in the past when doing a similar thing was to have no `#ifdef`s in header files, so that the interface was just one (as opposed to 2 or more versions based on how many ifdefs were there). Then the `#ifdef`s were in the .C where each section dealt with the one case. That way it was clear if the package was available or not and therefore what was safe to call... And than the clients of the interface did not have to know if the feature was available or not, they could just call the API...

fdkong commented:
@andrsd, thanks a lot for your time.
